File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int low=0,high=0, n=fruits.length , maxLen = 0; 
        Map<Integer, Integer> map = new HashMap<>();

        while(high<n){
            map.put(fruits[high] , map.getOrDefault(fruits[high],0)+1);

            if(map.size()<=2){
                maxLen = Math.max(maxLen , high-low+1);
            }
            else{
                while(map.size()>2){
                    map.put(fruits[low] , map.get(fruits[low])-1);
                    if(map.get(fruits[low])==0){
                        map.remove(fruits[low]);
                    }
                    low++;
                }
            }
            high++;
        }
        return maxLen ; 
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
from collections import defaultdict
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n=len(fruits)
        low=0
        high=0
        resLen=0
        fruit_Map = defaultdict(int)
        while high<n:
            fruit_Map[fruits[high]]+=1
            if len(fruit_Map)<=2:
                resLen = max(resLen, high-low+1)
                high+=1
            else:   #invalid window
                while len(fruit_Map)>2:    
                    fruit_Map[fruits[low]] -= 1
                    if fruit_Map[fruits[low]]==0:
                        del fruit_Map[fruits[low]]
                    low+=1
                high+=1   
        return resLen